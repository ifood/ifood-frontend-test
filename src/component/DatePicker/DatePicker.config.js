import { registerLocale } from 'react-datepicker';
import { ptBR, enUS } from 'date-fns/esm/locale';

registerLocale('pt-BR', ptBR);
registerLocale('en-US', enUS);
