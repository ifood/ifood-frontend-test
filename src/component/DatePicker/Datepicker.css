/* This is needed because react-datepicker does not suporte styling */
.datepicker--fluid {
  width: 100%;
}
