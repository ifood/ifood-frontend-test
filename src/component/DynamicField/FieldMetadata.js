import { formatDate, getLanguageName } from '../../core/l10n/L10n';

const SUPORTED_TYPE = {
  SELECT: Symbol('select'),
  DATE_TIME: Symbol('date_time'),
  NUMBER: Symbol('number'),
};

/**
 * TODO: This converter is not as dynamic as it should be
 */
const SPECIFIC_FIELD_CONVERTER = {
  locale: ({ name }) => getLanguageName(name),
};

const selectValuesMapper = ({ id, values }) => {
  return values.map((item) => {
    const formatter = SPECIFIC_FIELD_CONVERTER[id];
    let text = item.name;
    if (formatter != null) {
      const formatedText = formatter(item);
      if (formatedText != null) {
        text = formatedText;
      }
    }
    return {
      key: item.value,
      text,
      value: item.value,
    };
  });
};

const FIELD_FORMATTER = {
  [SUPORTED_TYPE.SELECT]: ({ fieldMetadata, value }) => {
    return selectValuesMapper(fieldMetadata).find(
      (item) => item?.value === value
    ).text;
  },
  [SUPORTED_TYPE.DATE_TIME]: ({ value }) =>
    value != null ? formatDate(value, 'Pp') : null,
};

const FIELD_SERIALIZER = {
  [SUPORTED_TYPE.DATE_TIME]: ({ value }) =>
    value != null ? value.toISOString() : null,
};

function getComponentType(metadata) {
  const { values, validation } = metadata;

  if (values != null) {
    return SUPORTED_TYPE.SELECT;
  }
  if (validation?.primitiveType === 'INTEGER') {
    return SUPORTED_TYPE.NUMBER;
  }
  if (validation?.entityType === 'DATE_TIME') {
    return SUPORTED_TYPE.DATE_TIME;
  }
  return null;
}

function getFieldFormatter(fieldMetadata) {
  const type = getComponentType(fieldMetadata);
  return FIELD_FORMATTER[type];
}

function getFieldSerializer(fieldMetadata) {
  const type = getComponentType(fieldMetadata);
  return FIELD_SERIALIZER[type];
}

function parseObjectByMetatada(metadata, object) {
  if (metadata == null || object == null) {
    return null;
  }
  return Object.keys(object).reduce((json, field) => {
    const fieldMeatada = metadata.find((item) => item.id === field);
    const serializer = getFieldSerializer(fieldMeatada);
    let value = object[field];
    if (serializer != null) {
      value = serializer({ value, fieldMeatada });
    }
    return {
      ...json,
      [field]: value,
    };
  }, {});
}

export {
  SUPORTED_TYPE,
  getFieldFormatter,
  getComponentType,
  parseObjectByMetatada,
  selectValuesMapper,
};
