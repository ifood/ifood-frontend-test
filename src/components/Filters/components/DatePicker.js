import React from 'react';
import { DatePicker } from 'antd';
import styled from 'styled-components';
import moment from 'moment';

const DatePickerContainer = styled.div`
  display: flex;
  flex-direction: column;
`;
const DatePickerLabel = styled.span`
  color: red;
`;
const DatePickertStyled = styled(DatePicker)`
  border-radius: 6px;
`;

export default function Page({ onChange = () => {}, ...data }) {
  const handleChange = (value) => {
    onChange(data.id, moment(value).format('YYYY-MM-DDTHH:MM:SSZ'));
  };

  return (
    <DatePickerContainer aria-label="spotifood-filter-date-time-container">
      <DatePickerLabel aria-label="spotifood-filter-date-time-text">
        Date Time
      </DatePickerLabel>
      <DatePickertStyled
        defaultValue={0}
        showTime
        onChange={handleChange}
        aria-label="spotifood-filter-date-time-picker"
      />
    </DatePickerContainer>
  );
}
