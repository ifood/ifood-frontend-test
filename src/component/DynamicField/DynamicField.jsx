import React from 'react';
import PropTypes from 'prop-types';
import { Select } from 'semantic-ui-react';
import DatePicker from 'react-datepicker';
import { useIntl } from 'react-intl';
import Field from '../Field';
import {
  getComponentType,
  SUPORTED_TYPE,
  selectValuesMapper,
} from './FieldMetadata';
import { getLocale } from '../../core/l10n/L10n';

function DynamicField({ metadata, value = '', onChange }) {
  const intl = useIntl();

  if (metadata == null) {
    return null;
  }

  function handleChange(data) {
    if (onChange != null) {
      onChange(data);
    }
  }

  const handleChangeSelect = (field) => (e, { value: newValue }) => {
    handleChange({ field, value: newValue });
  };

  const handleChangeInput = (field) => (e) => {
    handleChange({
      field,
      value: e?.target?.value,
    });
  };

  const handleChangeDate = (field) => (newValue) => {
    handleChange({ field, value: newValue });
  };

  function getComponentByMetadata(fieldMetadata) {
    const { values, name, id, validation } = fieldMetadata;
    const type = getComponentType(fieldMetadata);

    switch (type) {
      case SUPORTED_TYPE.SELECT:
        return [
          Select,
          {
            fluid: true,
            options: selectValuesMapper({ id, values }),
            placeholder: name,
            value,
            onChange: handleChangeSelect(id),
          },
        ];
      case SUPORTED_TYPE.NUMBER:
        return [
          'input',
          {
            id,
            placeholder: name,
            type: 'number',
            min: validation?.min,
            max: validation?.max,
            value,
            onChange: handleChangeInput(id),
          },
        ];
      case SUPORTED_TYPE.DATE_TIME:
        return [
          DatePicker,
          {
            id,
            wrapperClassName: 'datepicker--fluid',
            placeholderText: intl.formatMessage({ id: 'datetime.placeholder' }),
            selected: value,
            onChange: handleChangeDate(id),
            dateFormat: 'Pp',
            timeInputLabel: intl.formatMessage({ id: 'time' }),
            showTimeInput: true,
            locale: getLocale(),
          },
        ];
      default:
        return [];
    }
  }

  const [Component, componentProps] = getComponentByMetadata(metadata, value);

  if (Component != null) {
    return (
      <Field label={metadata.name} fieldId={metadata.id}>
        <Component {...componentProps} />
      </Field>
    );
  }
  return null;
}

DynamicField.propTypes = {
  metadata: PropTypes.object.isRequired,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.instanceOf(Date)]),
  onChange: PropTypes.func,
};

export default React.memo(DynamicField);
