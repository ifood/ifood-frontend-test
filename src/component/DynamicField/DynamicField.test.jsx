import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Select } from 'semantic-ui-react';
import DatePicker from 'react-datepicker';
import DynamicField from './DynamicField';

const numberMetadata = {
  id: 'limit',
  name: 'Quantidade',
  validation: {
    primitiveType: 'INTEGER',
    min: 1,
    max: 50,
  },
};

const selectMetadata = {
  id: 'locale',
  name: 'Locale',
  values: [
    {
      value: 'pt_BR',
      name: 'pt_BR',
    },
    {
      value: 'en_US',
      name: 'en_US',
    },
  ],
};

const dateTimeMetadata = {
  id: 'timestamp',
  name: 'Data e Hor√°rio',
  validation: {
    primitiveType: 'STRING',
    entityType: 'DATE_TIME',
    pattern: 'yyyy-MM-ddTHH:mm:ss',
  },
};

const notSupportedMetadata = {
  validation: {
    primitiveType: 'SOME_NOT_SUPPORTED_THING',
  },
};

describe('<DynamicField />', () => {
  it('snapshot', () => {
    const dynamicField = mount(<DynamicField metadata={numberMetadata} />);
    jestExpect(toJson(dynamicField)).toMatchSnapshot();
  });

  it('should return null if metadata is not provided', () => {
    const dynamicField = shallow(<DynamicField />);
    expect(dynamicField.type()).to.be.null;
  });

  describe('supported field types', () => {
    it('Select', () => {
      const dynamicField = shallow(<DynamicField metadata={selectMetadata} />);
      expect(dynamicField.find(Select)).to.have.lengthOf(1);
    });

    it('Number', () => {
      const dynamicField = shallow(<DynamicField metadata={numberMetadata} />);
      expect(dynamicField.find('input[type="number"]')).to.have.lengthOf(1);
    });

    it('DateTime', () => {
      const dynamicField = shallow(
        <DynamicField metadata={dateTimeMetadata} />
      );
      expect(dynamicField.find(DatePicker)).to.have.lengthOf(1);
    });

    it('Not supported type should return null', () => {
      const dynamicField = shallow(
        <DynamicField metadata={notSupportedMetadata} />
      );
      expect(dynamicField.type()).to.be.null;
    });
  });
});
