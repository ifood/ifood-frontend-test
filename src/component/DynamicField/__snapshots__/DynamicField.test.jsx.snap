// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DynamicField /> snapshot 1`] = `
<Memo(DynamicField)
  metadata={
    Object {
      "id": "limit",
      "name": "Quantidade",
      "validation": Object {
        "max": 50,
        "min": 1,
        "primitiveType": "INTEGER",
      },
    }
  }
>
  <Field
    fieldId="limit"
    label="Quantidade"
  >
    <FormField>
      <div
        className="field"
      >
        <label
          htmlFor="limit"
        >
          Quantidade
        </label>
        <input
          id="limit"
          max={50}
          min={1}
          onChange={[Function]}
          placeholder="Quantidade"
          type="number"
          value=""
        />
      </div>
    </FormField>
  </Field>
</Memo(DynamicField)>
`;
